posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,265 @@
+are parallelized - point-wise iterations can be freely parallelized along any dimension
+and grain size while reduction operations have to be either parallelized along dimensions
+that you're not iterating over or by performing bisect and reduce operations along the
+dimension being iterated. Parallelization can also happen using vectorized operations.
+
+## Iteration details
+
+The simplest iteration operation can be performed using the 
+[`for_each`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L525) 
+function. This function has two overloads - one which accepts a loop of type `loop_t`
+and another which accepts a `loop2d_t` (find them [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.h#L166)). The former can iterate over a loop
+of a single dimension whereas the latter can do so over two dimensions. The simplest
+way of using `loop_t` is to pass it as a lambda using to `for_each`. A code snippet
+using it this way would look like so:
+
+``` cpp
+auto iter = TensorIterator();
+iter.add_output(out);
+iter.add_input(a);
+iter.dont_resize_outputs(); // call if out is allocated.
+iter.dont_compute_common_dtype(); // call if inputs/outputs are of a different type.
+iter.build();
+
+auto loop = [&](char **data, const int64_t* strides, int64_t n) {
+    auto * out_data_bytes = data[0];
+    auto * in_data_bytes = data[1];
+    
+    // do something with input and output.
+    
+    out_data_bytes += strides[0];
+    in_data_bytes += strides[1];
+}
+
+iter.for_each(loop);
+```
+In the above example, the `char** data` gives a pointer to the data within the
+tensor in the same order that you specify when you build the iterator. Note
+that in order to make the implementation agnostic of any particular data type, you
+will always receive the data typecast to `char` (think of it as a bunch of bytes).
+
+The second argument is `int64_t* strides` which is an array containing the strides of
+the dimension that you're iterating over. We can add this stride to the pointer received
+in order to reach the next element in the tensor. The last argument is `int64_t n` which
+is the size of the dimension being iterated over. 
+
+The `for_each` loop will implicitly parallelize each iteration of `loop` if the size
+of each iteration is more than the value of `internal::GRAIN_SIZE`, which is a value
+that is determined as the 'right amount' of data to iterate over in order to gain a significant
+speedup using multi-threaded execution. If you want to explicitly specify that your
+operation _must_ run in serial, then use the `serial_for_each` loop.
+
+### Using kernels for iterations
+
+Frequently we want to iterate the elements of a tensor over the same kernel function.
+`TensorIterator`
+provides various such 'kernel iterators' that can be used for iterating over the elements
+of a tensor without having to worry about the stride, data type of the operands or details
+of the parallelism.
```

Comment:
Isn't this confusing kernels with point-wise operations? AFAIK the kernel applies to the whole tensor:

```suggestion
Frequently we want to create a kernel that applies a simple point-wise function onto entire tensors.
`TensorIterator`
provides various generic kernels that can be used for iterating over the elements
of a tensor without having to worry about the stride, data type of the operands or details
of the parallelism.
```

Reply:
Makes sense.